### ISSUE TITLE
MKV support

### ISSUE THREAD

**Issue by feross** [jhiesey/videostream#29]
Right now, WebTorrent uses the `videostream` package to provide support for MP4. It would be great to be able to stream MKV files since it's a very popular container format.

**Comment by v0dy** [comment]
This library might help! https://github.com/RReverser/mpegts

**Comment by lesander** [comment]
How about WebChimera, @feross? http://www.webchimera.org/

**Comment by feross** [comment]
WebChimera does not work in the browser. But it's an option for Electron or
nw.js apps.

**Comment by ma-tt** [comment]
+1 for .mkv support.

**Comment by OfertasDescontos** [comment]
+1 for .mkv support.

**Comment by feross** [comment]
@ma-tt @OfertasDescontos Please don't post useless +1 comments. Use the "reactions" feature to indicate your support for an idea. Thanks.

**Comment by qgustavor** [comment]
There is [node-matroska](https://github.com/oeuillot/node-matroska), "Matroska library written for nodejs".  While node-matroska isn't easily browserifyable the original project, [node-ebml](https://github.com/themasch/node-ebml), is. I used it to make [this tool](https://qgustavor.github.io/mkv-extract/). My code [is quite messy](https://github.com/qgustavor/mkv-extract/), consider it a proof-of-concept. Maybe if node-matroska gets improved and finished it can be used here.

**Comment by geecko86** [comment]
Any update on this front?

**Comment by HostFat** [comment]
Yep, it will be great :)

**Comment by norsegopher** [comment]
@jhiesey, please give us some of your magic and add MKV support.

**Comment by regalstreak** [comment]
>inb4 just use vlc

**Comment by Miaourt** [comment]
Hello, monthly silly up, if there is any tips on how we could help for this btw

**Comment by PavelShar** [comment]
Hello, guys!
Any updates about MKV stream support?

**Comment by zekiblue** [comment]
mkv support is essential guys, most of the videos are served in mkv

**Comment by qgustavor** [comment]
@zekiblue I don't think so: maybe most videos distributed illegally are served using this container (unless you also consider WebM, which is based on Matroska, but isn't MKV), Anyway I'm just guessing: streaming services, AFAIK, are huge, do not use MKV and count for "most of the videos served".

Matroska is just an container, the main issue on supporting MKV is that many videos I find using this container use codecs not supported by many browsers. You can test it by running `ffmpeg -i your-video.mkv -codec copy test.mp4` then trying to play the video in a browser. I just tried some random video and it failed.

That being said, if the codec issue gets ignored, one way to solve this issue is reproducing the above command using JavaScript: use [node-ebml](https://github.com/node-ebml/node-ebml) to parse the MKV file (it works using streams, which is ideal), extract streams from it and [remux those into MP4](https://github.com/samirkumardas/jmuxer). Subtitle streams can also be extracted and forwarded to some renderer such as JavascriptSubtitlesOctopus or Captionator.js.

**Comment by zisra** [comment]
Maybe we will have it in 2030.

**Comment by diegocr** [comment]
@zisra lol... 😉 

Well, if it does help... https://github.com/meganz/videostream/blob/master/ebml-remuxer.js

It just does lack the use of `WebCodecs` to handle all kind of video-codecs supported by browsers nowadays, PRs welcome :)

**Comment by endlands** [comment]
Chrome and Firefox both support MKV with H.264, AV1, AAC, Opus, and others by default without any feature flag. Not sure about Safari.

**Comment by diegocr** [comment]
@endlands I thought they do only support WebM through MSE with supported codecs, i.e. vp8, vp9, opus and av1, and that's what the above ebml-remuxer does, pretending MKVs to be a WebM it it's with either of those codecs -- it was ages ago last time i researched about this, tho - is that with recent browser versions? any further details are greatly welcome, thanks!
